examples/doc/aadc.1
.\" aadc.1 - modified from dc.1 by Jason McCarty on 2002-05-16
.\" --begin original dc.1 copyright
.\"
.\" dc.1 - the *roff document processor source for the dc manual
.\"
.\" This file is part of GNU dc.
.\" Copyright (C) 1994, 1997, 1998, 2000 Free Software Foundation, Inc.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License , or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to:
.\"   The Free Software Foundation, Inc.
.\"   59 Temple Place, Suite 330
.\"   Boston, MA 02111 USA
.\"
.\" --end original dc.1 copyright
.TH AADC 1 "2002-05-27" "Jason McCarty"
.ds dc \fIaadc\fP
.ds Dc \fIAadc\fP
.ds st \fISelf-Validated Numerical Methods and Applications\fP

.SH NAME
aadc \- a dc-like Affine Arithmetic calculator

.SH SYNOPSIS
aadc \" [-V] [--version] [-h] [--help]
.\"   [-e scriptexpression] [--expression=scriptexpression]
.\"   [-f scriptfile] [--file=scriptfile]
.\"   [file ...]

.SH DESCRIPTION
\*(Dc is a reverse-polish calculator implementing Affine Arithmetic
as described in \*(st (available from
\fIhttp://www.dcc.unicamp.br/~stolfi/EXPORT/projects/affine-arith/\fP
as of this writing). \*(Dc is a limited work-alike of the standard UNIX
program \fBdc\fP(1).

Numbers can be entered in standard decimal and scientific notation, with the
exception that negative numbers must use ``_'' (underscore) instead of
``-'' (hyphen) in front of the number.

An affine form is entered as ``<\fIcenter\fP, \fIradius\fP>,'' which creates
a new affine form centered at \fIcenter\fP, with radius \fIradius\fP.

Arguments and commands must be seperated by whitespace, since unlike \fIdc\fP,
\*(dc has multi-character commands.

.ig
.SH OPTIONS
\*(Dc may be invoked with the following command-line options:
.TP
.B -V
.TP
.B --version
Print out the version of \*(dc that is being run and a copyright notice,
then exit.
..
.ig
.TP
.B -h
.TP
.B --help
Print a usage message briefly summarizing these command-line options
and the bug-reporting address,
then exit.
.TP
.B -e \fIscript\fP
.TP
.BI --expression= script
Add the commands in
.I script
to the set of commands to be run while processing the input.
.TP
.B -f \fIscript-file\fP
.TP
.BI --file= script-file
Add the commands contained in the file
.I script-file
to the set of commands to be run while processing the input.
.PP
If any command-line parameters remain after processing the above,
these parameters are interpreted as the names of input files to
be processed.
A file name of
.B -
refers to the standard input stream.
The standard input will processed if no file names are specified.
.PD
..

.SH COMMANDS
.SS
Printing Commands
.TP
.B p
Prints the value on the top of the stack,
without altering the stack.
.TP
.B f
Prints the entire contents of the stack
.ig
and the contents of all of the registers,
..
without altering anything.
This is a good command to use if you are lost or want
to figure out what the effect of some command has been.
.TP
.B k
Pops the top value on the stack, and sets the display precision (number of
digits printed) to be the lower bound of the value popped, rounded to an
integer.
Note that this command does not affect \*(dc's internal accuracy, only the way
numbers are printed.
.TP
.B terse
Don't show deviations when printing output.
.TP
.B verbose
Show deviations when printing output (the default setting).
.PD
.SS
Arithmetic
.TP
.B +
Pops two values off the stack, adds them,
and pushes the result.
.TP
.B -
Pops two values,
subtracts the first one popped from the second one popped,
and pushes the result.
.TP
.B *
Pops two values, multiplies them, and pushes the result.
.TP
.B /
Pops two values,
divides the second one popped from the first one popped,
and pushes the result.
.TP
.B ^
Pops two values and exponentiates,
using the first value popped as the exponent
and the second popped as the base.

.SS "Additional Mathematical Functions"
.ds pp Pops one value, computes
.ds pu and pushes the result.
.TP
.B abs
\*(pp its absolute value, \*(pu
.TP
.B cos
\*(pp its cosine, \*(pu
.TP
.B exp
\*(pp \fIe\fP taken to that power, \*(pu
.TP
.B log
\*(pp its natural logarithm, \*(pu
.TP
.B rand
\*(pp a random number within its range, \*(pu
.TP
.B sin
\*(pp its sine, \*(pu
.TP
.B sincos
\*(pp its sine and cosine, and pushes them, in that order. \fBsincos\fP
is faster than ``d sin r cos'', so it is useful when both the sine and
cosine of the same value are desired. Furthermore, the values returned
will be correlated by a trigonometric identity, which will improve
accuracy in some calculations.
.TP
.B sqr
\*(pp its square, \*(pu
This is faster and more accurate than ``d *'' or ``2 ^'' since \*(dc knows
exactly what operation is desired.
.TP
.B sqrt
\*(pp its square root, \*(pu
.TP
.B tan
\*(pp its tangent, \*(pu
.TP
.B v
Alias for \fBsqrt\fP.

.SS Stack Control
.TP
.B c
Clears the stack, rendering it empty. This does not affect named variables.
.TP
.B d
Duplicates the value on the top of the stack and pushes another copy of it.
.TP
.B pop
Pops the value on the top of the stack, and discards it.
.TP
.B r
Reverses the order of (swaps) the top two values on the stack.

.SS "Affine Arithmetic Utility Operations"
.TP
.B affine
Pops the top two values off the stack, creates a new affine form centered at
the lower bound of the second value popped, with radius equal to the lower
bound of the first value popped, then pushes that form. Thus, ``1 2 affine''
has the same effect as ``<1, 2>''.
.TP
.B collapse
The value on top of the stack is ``collapsed.'' This means that all the
deviations in this value but not present in any other value are merged into a
single deviation. This can improve speed and memory usage, and reduce the
amount of output when it is printed.

.SS Variables
Instead of single-letter registers, \*(dc has named variables. A variable
name can be any string, so long as it doesn't contain any whitespace
characters. You could remember the function of \fB$\fP and \fB!\fP by
thinking of them as \fBs\fP and \fBl\fP from \fIdc\fP, respectively.
.TP
\fB$\fP\fIfoo\fP
Pops one value from the stack and stores it in the named variable \fIfoo\fP.
.TP
\fB!\fP\fIfoo\fP
Pushes the value of \fIfoo\fP onto the stack, without modifying \fIfoo\fP.
.TP
\fB#\fP\fIfoo\fP
Removes \fIfoo\fP from the symbol table; it is no longer recognized as a
variable. A good mnemonic for this is that you ``pound foo'' to delete it.

.SS Miscellaneous
.TP
.B q
Quit \*(dc.
.TP
.B z
Pushes the current stack depth:
the number of objects on the stack before the execution of the
.B z
command.

.SH EXAMPLES
.\" FIXME: write more.
The following line will determine the range of the function
y = x / sin x, when x is in the range [4..5], and save the values
of x and y for future use.
.RS
<4.5, 0.5> d d sin / p $y $x
.RE

.SH "SIGNIFICANT DIFFERENCES BETWEEN AADC AND DC"
Because \*(dc allows multi-character commands, all commands and arguments must
be seperated by whitespace.

\*(Dc does not implement macros or string handling of any kind.

\*(Dc has variables instead of registers, but unlike \fIdc\fP, \*(dc's
variables are not stacks or arrays, just single values.

\*(Dc does not allow you to set the precision or radixes. It uses
double-precision floating point internally.

The \fB!\fP operator in \*(dc does not execute system commands as in \fIdc\fP.
It pushes a variable onto the stack instead.

Other differences (mostly features in \fIdc\fP but not in \*(dc) can be
determined by contrasting this manual page with \fBdc\fP(1).

.SH "SEE ALSO"
.BR dc (1)
and \*(st, L. H. de Figueiredo and J. Stolfi

.SH BUGS
Many more functions need to be implemented and refined, especially trig
functions.

Using ``_'' for negative numbers has a quirk: typing ``_-5'', for example, will
enter a positive five, since the input routines negate it twice.

\*(Dc does not yet have a license.

.SH AUTHOR
.nf
Jason McCarty <bclg@iup.edu>
.fi

.SH ACKNOWLEDGEMENTS
Thanks are due to Jorge Stolfi and Luiz Henrique de Figueiredo for
inventing/discovering Affine Arithmetic and distributing the details in
writing. Thanks are also due to the Free Software Foundation for their
implementation of \fIdc\fP, which introduced me to reverse-polish notation, and
for the many other UNIX utilities written by them. The groff source of this
manual page is based on \fIdc\fP's manual page, along with several other
utilities' manuals.


examples/aadc.cpp
#include <iostream>
#include <utility>
#include <iomanip>
#include <deque>
#include <map>
#include <string>
#include <cstdlib>
#include <ctime>
#include <affine/affine.h>

#ifdef HAVE_CONFIG_H
#  include <config.h>
#  define VERSION " " PACKAGE_VERSION REVISION " (Built " DATE ")"
#else
#  define VERSION ""
#endif

using std::map;
using std::string;
typedef Affine OperandType;
typedef std::deque<OperandType> OperandStack;
typedef std::pair<void (*) (OperandStack&), unsigned int> Operator;
//                function to call         number of args

typedef map<string, Operator> OperatorLUT; //operator look-up table
void init_operators (OperatorLUT&);
void call_operator (const OperatorLUT::const_iterator&, OperandStack&);

class except {};

inline void bad_input (const string& s)
{
  std::cerr << s;
  throw except();
}

int main (int argc, char** argv)
{
  using std::cin;
  using std::cerr;
  OperandStack operands;
  map<string, OperandType> variables;
  variables["pi"] = IntervalH::pi();
  variables["e"] = IntervalH::e();

  OperatorLUT operators;
  init_operators(operators);

  std::srand(std::time(0));

  string malformed("malformed input");

  if (argc <= 1 || strcmp(argv[1], "-q"))
    std::cout << "Aadc" VERSION << "\nCopyright 2002-2003 Jason McCarty\n";

  char firstchar = 0;
  while (cin >> firstchar)
  { try {
    cin.putback(firstchar);
    if (firstchar == '<')
    {
      operands.push_front(Affine());
      if (!(cin >> operands.front())) {
        operands.pop_front();
        bad_input(malformed);
      }
    }
    else if (firstchar == '[')
    {
      Interval temp;
      if (cin >> temp)
        operands.push_front(Affine(temp));
      else
        bad_input(malformed);
    }
    else if ((firstchar >= '0' && firstchar <= '9') || firstchar == '.'
        || firstchar == '_')
    {
      if (firstchar == '_')
        cin >> firstchar;
      double temp;
      if (cin >> temp)
        operands.push_front(Affine((firstchar == '_') ? -temp : temp));
      else
        bad_input(malformed);
    }
    else if (firstchar == '$') //pop into a variable
    {
      if (!operands.size())
        bad_input("stack underflow");
      char scratch;
      string temp;
      if(cin >> scratch >> temp)
      {
        variables[temp] = operands.front();
        operands.pop_front();
      }
      else
        bad_input(malformed);
    }
    else if (firstchar == '!') //push from a variable
    {
      char scratch;
      string temp;
      if(cin >> scratch >> temp)
      {
        map<string, OperandType>::const_iterator p
          = variables.find(temp);
        if (p != variables.end())
          operands.push_front(variables[temp]);
        else {
          cerr << temp;
          bad_input(" is not an initialized variable");
        }
      }
    }
    else if (firstchar == '#') //delete a variable
    {
      char scratch;
      string temp;
      if(cin >> scratch >> temp)
      {
        map<string, OperandType>::iterator p
          = variables.find(temp);
        if (p != variables.end())
          variables.erase(p);
        else {
          cerr << temp;
          bad_input(" is not an initialized variable");
        }
      }
    }
    else
    {
      string temp;
      if (cin >> temp)
      {
        const OperatorLUT::const_iterator p = operators.find(temp);
        if (p == operators.end()) {
          cerr << temp;
          bad_input(" is not a valid operator");
        }
        else {
          try {
            call_operator(p, operands);
          }
          catch (Affine::DomainError)
          {
            cerr << temp;
            bad_input(": domain error");
          }
        }
      }
    }
  }
  catch (except)
  {
    if (cin.eof())
      cerr << ". EOF received\n";
    else {
        string temp("");
        cin.clear();
        std::getline(cin, temp);
        cerr << ". Location: ";
        if (temp.size())
          cerr << "at or before `" << temp;
        else
          cerr << "at or after `" << firstchar;
        cerr << "'. Remainder of line discarded.\n";
      }
    }
  }
  return 0;
}

void collapse (OperandStack& operands)
{ operands.front().collapse(); }

void clear (OperandStack& operands)
{ operands.clear(); }

void dup (OperandStack& operands)
{ operands.push_front(operands.front()); }

void printall (OperandStack& operands)
{
  for (OperandStack::const_iterator i = operands.begin();
      i != operands.end(); ++i)
    std::cout << *i << "\n";
}

void printtop (OperandStack& operands)
{ std::cout << operands.front() << "\n"; }

void quit (OperandStack&)
{ std::exit(0); }

void pushdepth (OperandStack& operands)
{ operands.push_front(OperandType(operands.size())); }

void setterse (OperandStack&)
{ std::cout << AffineH::terse; }

void setverbose (OperandStack&)
{ std::cout << AffineH::verbose; }

void abs (OperandStack& operands)
{ absA(operands.front()); }

void cos (OperandStack& operands)
{ cosA(operands.front()); }

void asin (OperandStack& operands)
{ asinA(operands.front()); }

void acos (OperandStack& operands)
{ acosA(operands.front()); }

void atan (OperandStack& operands)
{ atanA(operands.front()); }

void exp (OperandStack& operands)
{ expA(operands.front()); }

void setprecision (OperandStack& operands)
{
  std::cout << std::setprecision(static_cast<int> (operands.front().lo()));
  operands.pop_front();
}

void log (OperandStack& operands)
{ logA(operands.front()); }

void pop (OperandStack& operands)
{ operands.pop_front(); }

void rand (OperandStack& operands)
{
  Affine& f = operands.front();
  f = Affine(f.lo() + std::rand() * width(f) / RAND_MAX);
}

void sin (OperandStack& operands)
{ sinA(operands.front()); }

void sincos (OperandStack& operands)
{
  operands.push_front(Affine());
  sincos(operands[1], &operands[1], &operands.front());
}

void sqr (OperandStack& operands)
{ sqrA(operands.front()); }

void sqrt (OperandStack& operands)
{ sqrtA(operands.front()); }

void tan (OperandStack& operands)
{ tanA(operands.front()); }

void add (OperandStack& operands)
{
  operands[1] += operands.front();
  operands.pop_front();
} 

void sub (OperandStack& operands)
{
  operands[1] -= operands.front();
  operands.pop_front();
} 

void mul (OperandStack& operands)
{
  operands[1] *= operands.front();
  operands.pop_front();
} 

void div (OperandStack& operands)
{
  operands[1] /= operands.front();
  operands.pop_front();
} 

void pow (OperandStack& operands)
{
  powA(operands[1], operands.front());
  operands.pop_front();
} 

void swap (OperandStack& operands)
{
  Affine temp(operands[1]);
  operands[1] = operands.front();
  operands.front() = temp;
}

void affine (OperandStack& operands)
{
  operands[1] = Affine(operands[1].lo(), operands.front().lo());
  operands.pop_front();
}

void call_operator (const OperatorLUT::const_iterator& function,
    OperandStack& operands)
{
  if (operands.size() < function->second.second)
    bad_input(function->first + ": stack underflow");
  else
    function->second.first(operands);
}

inline Operator make_pair (void (*f) (OperandStack&), unsigned int n)
{ return std::make_pair<void (*) (OperandStack&), unsigned int>(f, n); }

void init_operators (OperatorLUT& operators)
{
  //operators which take 0 operands
  operators["c"]        = make_pair(clear, 0);
  operators["collapse"] = make_pair(collapse, 0);
  operators["f"]        = make_pair(printall, 0);
  operators["q"]        = make_pair(quit, 0);
  operators["verbose"]  = make_pair(setverbose, 0);
  operators["terse"]    = make_pair(setterse, 0);
  operators["z"]        = make_pair(pushdepth, 0);

  //operators which take 1 operand
  operators["abs"]    = make_pair(abs, 1);
  operators["acos"]   = make_pair(acos, 1);
  operators["asin"]   = make_pair(asin, 1);
  operators["atan"]   = make_pair(atan, 1);
  operators["cos"]    = make_pair(cos, 1);
  operators["d"]      = make_pair(dup, 1);
  operators["exp"]    = make_pair(exp, 1);
  operators["k"]      = make_pair(setprecision, 1);
  operators["log"]    = make_pair(log, 1);
  operators["p"]      = make_pair(printtop, 1);
  operators["pop"]    = make_pair(pop, 1);
  operators["rand"]   = make_pair(rand, 1);
  operators["sin"]    = make_pair(sin, 1);
  operators["sincos"] = make_pair(sincos, 1);
  operators["sqr"]    = make_pair(sqr, 1);
  operators["sqrt"]   = make_pair(sqrt, 1);
  operators["tan"]    = make_pair(tan, 1);
  operators["v"]      = make_pair(sqrt, 1);

  //operators which take 2 operands
  operators["+"] = make_pair(add, 2);
  operators["-"] = make_pair(sub, 2);
  operators["*"] = make_pair(mul, 2);
  operators["/"] = make_pair(div, 2);
  operators["^"] = make_pair(pow, 2);
  operators["r"] = make_pair(swap, 2);
  operators["affine"] = make_pair(affine, 2); //make an affine form from top two
}
